### PR TITLE
test: introduce `libp2p-swarm-test`

### LINKED ISSUE #2884: Introduce `libp2p-swarm-test` crate
## Description

Introduce a crate that contains reusable bits that make it easy to correctly test `NetworkBehaviour` implemenations. In particular, the crate should support the setting up and connecting a number of swarms, driving them correctly and asserting the emitted events.

## Motivation

For code to be high quality it should be tested. If testing is easier, users are more likely to do it, thus yielding higher quality code.

## Are you planning to do it yourself in a pull request?

Yes.

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

This patch-set introduces `libp2p-swarm-test`. It provides utilities for quick and safe bootstrapping of tests for `NetworkBehaviour`s. The main design features are:

- Everything has timeouts
- APIs don't get in your way
- Minimal boilerplate

Closes #2884.

## Notes

I did not port `libp2p-kad` because the testing logic was too complicated. I also did not add tests for protocols that don't already have some. That is out of scope for this PR.

## Open Questions

- ~Should this testing crate also pull together utilities like [`ArbitraryPeerId`](https://github.com/kpp/rust-libp2p/blob/bc77431620f952dd8a9e0054bba73bb1c9638bb7/protocols/kad/src/query/peers/closest/disjoint.rs#L535-L546)?~
	- No. That should go into `quickcheck-ext`.
- ~Can we make the `Swarm::new_ephemeral` API any better? At the moment it only allows to configure the underlying transport but the authentication and muxer are hardcoded. I don't want to introduce any more parameters because it will get too verbose and there wouldn't be a benefit in having the function in the first place. Should we perhaps introduce a dedicated `SwarmBuilder` for our tests that start with some reasonable test defaults, like a `MemoryTransport`?~
- ~Trying to port the new harness to `libp2p-ping`, I noticed that we'd lose the `MuxerChoice` as part of the prop test. If we go forward with the `SwarmBuilder` extension idea, we could make a function there that cycles through various muxers and authentication algorithms to make sure things work across various stacks. I am not sure it is worth it though to be honest.~

## Open tasks

- ~Port to more crates~

## Change checklist

- [x] I have performed a self-review of my own code
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
~- [ ] A changelog entry has been made in the appropriate crates~

### CODE REVIEW

**Review by melekes — COMMENTED** [review]
I'm in great support of this ❤️

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs]:
```suggestion
    /// Listens on a random tcp address, polling the [`Swarm`] until the transport is ready to accept connections.
```

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs]: Having to provide a timeout seems off to me. How would a user decide on the right value? Can't we just rely on the overall test timeout?

File: swarm/src/lib.rs
Comment (mxinden) [inline swarm/src/lib.rs:302]:
```suggestion
    pub fn try_into_behaviour_event(self) -> Result<TBehaviourOutEvent, Self> {
```

Neat return signature!

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs]:
```suggestion
    async fn connect<T>(&mut self, other: &mut Swarm<T>)
```

- `block_on` reminds me of `futures::executor::block_on` which this doesn't use.
- `block_on_connection` does not imply actually triggering the dial

What do you think of the new name?

**Review by mxinden — COMMENTED** [review]
Direction looks good to me. Thanks for pushing this forward.

**Comment by mxinden** [comment]
> Should this testing crate also pull together utilities like [`ArbitraryPeerId`](https://github.com/kpp/rust-libp2p/blob/bc77431620f952dd8a9e0054bba73bb1c9638bb7/protocols/kad/src/query/peers/closest/disjoint.rs#L535-L546)?

I am guessing that defining `Arbitrary` on `PeerId` directly under the `cfg[test]` doesn't expose outside of the crate. Is that correct?

**Comment by mxinden** [comment]
> * Can we make the `Swarm::new_ephemeral` API any better? At the moment it only allows to configure the underlying transport but the authentication and muxer are hardcoded. I don't want to introduce any more parameters because it will get too verbose and there wouldn't be a benefit in having the function in the first place. Should we perhaps introduce a dedicated `SwarmBuilder` for our tests that start with some reasonable test defaults, like a `MemoryTransport`?

I think all our non `Transport` tests should use the `MemoryTransport`, thus I am in favor of this default.

**Comment by mxinden** [comment]
> Trying to port the new harness to `libp2p-ping`, I noticed that we'd lose the `MuxerChoice` as part of the prop test. If we go forward with the `SwarmBuilder` extension idea, we could make a function there that cycles through various muxers and authentication algorithms to make sure things work across various stacks. I am not sure it is worth it though to be honest.

I still have to put more thoughts into this, though as an intuition I suggest that a test only tests one thing. E.g. the `libp2p-ping` tests test `libp2p-ping` and not some permutation of different transports and muxers. I think that should happen in a separate dedicated test.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs]: Tests don't have an overall timeout unless the user wraps it all in one which is boilerplate and thus doesn't happen very often. Plus, compared to a test timeout, having the test failure in here provides a nice stacktrace that is easier to debug on where the test is failing.

The reason we can't have a timeout default is because we don't know about the test case, only the dev writing the test can decide that :)

File: swarm/src/lib.rs
Comment (thomaseizinger) [reply swarm/src/lib.rs:302]: :pray:

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs]: Yeah sounds good and it is shorter too :)

**Comment by thomaseizinger** [comment]
> > Should this testing crate also pull together utilities like [`ArbitraryPeerId`](https://github.com/kpp/rust-libp2p/blob/bc77431620f952dd8a9e0054bba73bb1c9638bb7/protocols/kad/src/query/peers/closest/disjoint.rs#L535-L546)?
> 
> I am guessing that defining `Arbitrary` on `PeerId` directly under the `cfg[test]` doesn't expose outside of the crate. Is that correct?

That is correct.

> > * Can we make the `Swarm::new_ephemeral` API any better? At the moment it only allows to configure the underlying transport but the authentication and muxer are hardcoded. I don't want to introduce any more parameters because it will get too verbose and there wouldn't be a benefit in having the function in the first place. Should we perhaps introduce a dedicated `SwarmBuilder` for our tests that start with some reasonable test defaults, like a `MemoryTransport`?
> 
> I think all our non `Transport` tests should use the `MemoryTransport`, thus I am in favor of this default.

Cool, that makes this simpler :)

> > Trying to port the new harness to `libp2p-ping`, I noticed that we'd lose the `MuxerChoice` as part of the prop test. If we go forward with the `SwarmBuilder` extension idea, we could make a function there that cycles through various muxers and authentication algorithms to make sure things work across various stacks. I am not sure it is worth it though to be honest.
> 
> I still have to put more thoughts into this, though as an intuition I suggest that a test only tests one thing. E.g. the `libp2p-ping` tests test `libp2p-ping` and not some permutation of different transports and muxers. I think that should happen in a separate dedicated test.

Yeah I am tending towards the same. We should have a good muxer test harness instead!

File: swarm-test/src/lib.rs
Comment (mxinden) [reply swarm-test/src/lib.rs]: > Tests don't have an overall timeout unless the user wraps it all in one which is boilerplate and thus doesn't happen very often. Plus, compared to a test timeout, having the test failure in here provides a nice stacktrace that is easier to debug on where the test is failing.

I wasn't aware of this. Thanks.

Still, I would consider concrete timeouts in tests a code smell. Users of this framework would have to make assumptions on the environment this is run on. E.g. they need to take into account that their test might run on a GitHub CI or some high performant developer machine.

I can not think of a way without the timeout while still having a strack trace. Is the strack trace important? I guess this boils down to "should one be able to see the root cause purely from the unit test failure, or do we expect a red test to need debugging anyways".

I don't expect any unit test to take longer than 10 seconds. In case we want to keep the stack traces, how about setting the timeout to a fixed `10s`?

**Comment by mxinden** [comment]
> > > Should this testing crate also pull together utilities like [`ArbitraryPeerId`](https://github.com/kpp/rust-libp2p/blob/bc77431620f952dd8a9e0054bba73bb1c9638bb7/protocols/kad/src/query/peers/closest/disjoint.rs#L535-L546)?
> > 
> > 
> > I am guessing that defining `Arbitrary` on `PeerId` directly under the `cfg[test]` doesn't expose outside of the crate. Is that correct?
> 
> That is correct.

That is unfortunate. In that case, I am in favor of providing an `ArbitraryPeerId` from this new crate.

File: swarm-test/src/lib.rs
Comment (elenaf9) [inline swarm-test/src/lib.rs]:
```suggestion
    Protocol::Memory(0).into()
```
The memory transport picks a random port if the port is 0.

File: swarm-test/src/lib.rs
Comment (elenaf9) [inline swarm-test/src/lib.rs]:
```suggestion
    fn new_ephemeral(behaviour_fn: impl FnOnce(Keypair) -> Self::NB) -> Self
```
Since the `PeerId` can be derived from the `Keypair`, what do you think about removing it as parameter?

File: swarm/src/lib.rs
Comment (elenaf9) [inline swarm/src/lib.rs:302]: Is there a case where a user would call this method and not unwrap the result? If not, what do you think about panicking here instead of returning an Err?

**Review by elenaf9 — APPROVED** [review]
Nice!

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs]: > I don't expect any unit test to take longer than 10 seconds. In case we want to keep the stack traces, how about setting the timeout to a fixed `10s`?

I guess for any situation where the timeout doesn't fit, the user can always fall back to the `StreamExt::next` function which doesn't have a timeout so this is probably a good idea?

What we need to balance with these test abstractions is for them to flexible enough to handle a lot of situations but also providing enough value in making test code concise. I think a timeout 10s is a good default and if it doesn't fit, you can always just code it yourself and it is not too verbose. If it happens a lot, we can look into adjusting the timeout or finding other solutions.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs]: That is so much clearer, thank you!

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs]: We could actually, yeah. It is a bit verbose though and many times, behaviours only need to know the `PeerId` and not the `Keypair`. Perhaps I can make some clever abstractions here so the user can choose ...

File: swarm/src/lib.rs
Comment (thomaseizinger) [reply swarm/src/lib.rs:302]: Given that this is a function in production code, I don't think panicking is a good option here.

Comment (elenaf9) [reply swarm/src/lib.rs:302]: Woops, I overlooked that it is in `swarm` and not `swarm-test`. Agree with you.

**Comment by thomaseizinger** [comment]
> > > > Should this testing crate also pull together utilities like [`ArbitraryPeerId`](https://github.com/kpp/rust-libp2p/blob/bc77431620f952dd8a9e0054bba73bb1c9638bb7/protocols/kad/src/query/peers/closest/disjoint.rs#L535-L546)?
> > > 
> > > 
> > > I am guessing that defining `Arbitrary` on `PeerId` directly under the `cfg[test]` doesn't expose outside of the crate. Is that correct?
> > 
> > 
> > That is correct.
> 
> That is unfortunate. In that case, I am in favor of providing an `ArbitraryPeerId` from this new crate.

Now that we have `quickcheck-ext`, should it perhaps go in there?

**Comment by mxinden** [comment]
@thomaseizinger let me know once I should be taking another look here.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs]: Addressed in [`c6b670d` (#2888)](https://github.com/libp2p/rust-libp2p/pull/2888/commits/c6b670d19b4a4c27cab660bc12688aa6578dd69c).

**Comment by thomaseizinger** [comment]
@libp2p/rust-libp2p-maintainers Please have another look to make sure this goes into the right direction! :)

**Comment by mxinden** [comment]
> @libp2p/rust-libp2p-maintainers Please have another look to make sure this goes into the right direction! :)

Right direction from my side.

**Comment by mxinden** [comment]
Still interested to get this merged @thomaseizinger! Writing a tiny protocol right now where this would come in handy for testing.

**Comment by thomaseizinger** [comment]
> Still interested to get this merged @thomaseizinger! Writing a tiny protocol right now where this would come in handy for testing.

Ah yes, me too. I'd prefer if we port the entire workspace over to this though which means there is still quite a bit of work to be done before we can merge it.

**Comment by thomaseizinger** [comment]
I'll see to spend some cycles on it soon.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [inline swarm-test/src/lib.rs:120]: I spent a good two hours of debugging timeouts in the test and unwoken tasks until it dawned on me why things were not working. @elenaf9 I believe this might also be the reason for your timeouts in #3345.

This function may seem like a beast but after trying various solutions locally, I am convinced it makes sense to add this to the test-harness and use this instead of `future::join` for testing `Swarm`s.

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs:21]: Please add a license header here :)

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs:108]: :exploding_head:

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs:160]: Long term it would be nice to detect missed wake-ups in a `NetworkBehaviour` implementation with `libp2p-swarm-test`. If I am not mistaken, that is not possible with the above `while` `select!` loop as long as the other swarm triggers a wake-up.

Lets defer to a future pull request.

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs]:
```suggestion
    /// The swarm will use a [`MemoryTransport`] together with [`Plaintext2Config`] authentication layer and
```

If I am not mistaken, this framework doesn't actually use noise.

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs:314]: :+1: good idea

File: swarm-test/src/lib.rs
Comment (mxinden) [inline swarm-test/src/lib.rs:211]: Why do we need the TCP transport? Is this needed for `libp2p-autonat`?

File: protocols/request-response/tests/ping.rs
Comment (mxinden) [inline protocols/request-response/tests/ping.rs:276]: Can this not be done via the new shiny `drive` function?

**Review by mxinden — APPROVED** [review]
Just looking at the diff, this is already a great step forward.

A couple of comments. Nothing major. Good to merge from my end.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs:314]: Thanks! I ended up doing it this way because it allowed to create the `connect` abstraction which can then rely on retrieving the addresses via the external addresses.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs:211]: Yes, and `dcutr` needs it too.

File: protocols/request-response/tests/ping.rs
Comment (thomaseizinger) [reply protocols/request-response/tests/ping.rs:276]: Probably, I've introduced that one only at the very end!

Comment (thomaseizinger) [reply protocols/request-response/tests/ping.rs:276]: I am not sure actually, this one is a bit different because the two peers operate autonomously. `drive` is useful if you want to "step" through a test in a more linear fashion because it gives you access to the `Swarm`s inbetween. Here, we can just create to tasks that operate independently.

File: swarm-test/src/lib.rs
Comment (thomaseizinger) [reply swarm-test/src/lib.rs:160]: This is kind of a conflicting requirement. I added this because it is very easy to poll a `Swarm` in a wrong way in a test, thus it is often the _test_ that is buggy, not the `NetworkBehaviour` implementation.

**Pull request merged**
